### ISSUE TITLE
ActionCable sporadically fails to confirm subscribe because the subscription "already exists"

### ISSUE THREAD

**Issue by shoffing** [rails/rails#48292]
### Steps to reproduce

- Create a very basic channel that simply streams for the current user:

```rb
class PerfTestChannel < ApplicationCable::Channel
  def subscribed
    stream_for current_user
  end
end
```

- On page load, subscribe to the channel.

```js
App.cable.subscriptions.create('PerfTestChannel');
```

### Expected behavior

The ActionCable client should be able to reliably subscribe to a channel.

### Actual behavior

- On a development instance with browser network latency simulation, or any staging/production instance, the `subscribe` command will sporadically (varying from 5%-70% depending on load) fail to be confirmed. The `SubscriptionGuarantor` will repeatedly attempt to resubscribe every 500ms, but no confirmation will ever be received.

![image](https://github.com/rails/rails/assets/2903742/8dc591b4-ef3a-4d7b-867d-b92ee6f5b906)

- I wrote a JS script to assist in reproducing this error. It repeatedly disconnects & subscribes to the channel until the subscription fails to leave the pendingSubscriptions state in under 600ms. Running this in browser JS console reliably reproduces the issue on a staging/production instance:

```js
(async function() {
  let i = 0;
  while (true) {
    i++;
    App.cable.subscriptions.subscriptions.forEach(sub => App.cable.subscriptions.remove(sub));
    App.cable.disconnect();
    App.cable.subscriptions.create('PerfTestChannel');
    await new Promise(resolve => setTimeout(resolve, 600));
    const pendingSubs = App.cable.subscriptions.guarantor.pendingSubscriptions;
    if (pendingSubs.length > 0) {
      break;
    } else { console.log('...' + i); }
  }
})()
```

### Investigation

This bug was tricky to investigate as there was nothing useful in our logs. The `/cable` request was successfully being upgraded to websockets as expected. We saw some errors that might have been relevant, but none of them occurred reliably with the issue being observed. After digging through the ActionCable source code, we have determined the issue is caused by this line in `subscriptions.rb`: https://github.com/rails/rails/blob/0f5c8c5bc61b7e382e64cad4846406021bc8cd35/actioncable/lib/action_cable/connection/subscriptions.rb#L34

It returns silently if the subscription "already exists". We confirmed this is the source of the error by monkey-patching the class in a `config/initializers` to add some logging with the `return`:

```rb
# config/initializers/monkeypatch_actioncable.rb
# frozen_string_literal: true

require "active_support/core_ext/hash/indifferent_access"

module ActionCable
  module Connection
    # Collection class for all the channel subscriptions established on a given connection. Responsible for routing incoming commands that arrive on
    # the connection to the proper channel.
    class Subscriptions # :nodoc:
      def add(data)
        id_key = data["identifier"]
        id_options = ActiveSupport::JSON.decode(id_key).with_indifferent_access

        if subscriptions.key?(id_key)
          Rails.logger.info("[Action Cable] Already subscribed to #{id_key}...")
          return
        end

        subscription_klass = id_options[:channel].safe_constantize

        if subscription_klass && ActionCable::Channel::Base > subscription_klass
          subscription = subscription_klass.new(connection, id_key, id_options)
          subscriptions[id_key] = subscription
          subscription.subscribe_to_channel
        else
          logger.error "Subscription class not found: #{id_options[:channel].inspect}"
        end
      end
    end
  end
end
```

We deployed the patch to staging and verified that when the error occurs and `subscribe` is being repeatedly attempted, our logs are filled with the "Already subscribed..." message. This appears to confirms that it's not an issue with our load balancer or nginx or something, the `subscribe` command is successfully making its way all the way to the Rails server. It's just being ignored, and failing silently.

### Possible Fix

Simply calling `subscribe_to_channel` if the subscription already exists seems to fix the issue:

```rb
return subscriptions[id_key].subscribe_to_channel if subscriptions.key?(id_key)
```

However, I'm not familiar enough with ActionCable to fully understand the implications of that logic. For example, I think that may result in multiple `subscribed` callbacks being invoked in some cases where multiple `subscribe` commands are sent from the client. There also seems to be some counting-semaphore logic wrapped around the connection to redis, which multiple subscribes could maybe interfere with.

### Impact

This bug is blocking our team's adoption of ActionCable. We cannot rely on the system if it fails to connect with even the most basic of setups.

### System configuration

**Rails version**: 7.0.4.3

**Ruby version**: 3.2.1

**Comment by shoffing** [comment]
We're still having severe trouble with this. In a recent test, this bug is causing a _**60% connection failure rate** on production traffic_! In a 10-minute window, there were 2100 successes and 3200 failures (with a failure being defined as the connection still pending after 5 seconds).

We seem to have tracked down the issue to the Redis pubsub subscribe callback not triggering. E.g., this line is never reached when this bug occurs:
https://github.com/rails/rails/blob/main/actioncable/lib/action_cable/channel/streams.rb#L91

We cannot use the Postgres adapter because our application uses MySQL. And, of course, the async adaptor is no use for production.

We thought the solution from this issue would help, but unfortunately the issue still persists: https://github.com/rails/rails/issues/27659

@spinosa this bug appears to be something your SubscriptionGuarantor was supposed to prevent (https://github.com/rails/rails/pull/41581), but it's not working in this instance. Is there anything you can think of to help resolve this? Have you seen anything like this before?

**Comment by shoffing** [comment]
Update: We gave up and switched to AnyCable. https://docs.anycable.io/rails/getting_started

We're at 100% traffic and seeing virtually zero connection failures now. This required spinning up extra infrastructure for the anycable-go server.

I still strongly believe this is an bug in ActionCable's source code, and that it's caused by a failure to subscribe to the Redis channels. ActionCable uses one redis channel per connection, while AnyCable uses a single Redis channel with messages containing the connection identifiers...maybe that has something to do with it?

Either way, I would recommend against using ActionCable alone for any production applications with more than a couple dozen users.

**Comment by philipithomas** [comment]
I'm hitting this same issue. 
<img width="518" alt="Screenshot 2023-08-24 at 09 13 39" src="https://github.com/rails/rails/assets/1312414/de1f08bd-0576-4394-9ae6-1113d98548db">

**Comment by tenpaiyomi** [comment]
Running across the same thing myself with no real luck in identifying the underlying cause.

**Comment by palkan** [comment]
Hey everyone,

> I still strongly believe this is a bug in ActionCable's source code, and that it's caused by a failure to subscribe to the Redis channels

I'm not sure Redis is involved. The code around Redis Pub/Sub is _protected_ by mutexes, so it's concurrent-safe. The `@subscriber_callbacks` are always executed on Redis subscription, even if we already requested `unsubscribe`.

The only place in code that I found suspicious is this: https://github.com/rails/rails/blob/bdbc1825c0b1c55033b2c9afa93a3c4987126442/actioncable/lib/action_cable/channel/streams.rb#L87-L95

We store stream in the `streams` Hash before confirming the subscription (or even executed—it's happening async). At the same time, we may trigger `#stop_all_streams` before the async callback (`pubsub.subscribe(...)`) is invoked and try to unsubscribe from the stream we haven't really subscribed to yet. That would lead to Redis `UNSUBSCRIBE stream` to be called before `SUBSCRIBE stream`, which is not a problem—we still send the confirmation (however, the client may no longer be interested in it—that's a different problem).

I think, repeated `subscribe` commands is a bug in `SubscriptionGuarantor` (or, more broadly, the current logic around dealing with concurrent subscriptions at the client-side). That's why I would recommend switching to [anycable-client](https://github.com/anycable/anycable-client) (and [@anycable/turbo-stream](https://github.com/anycable/anycable-client/tree/master/packages/turbo-stream) respectively) and forget about this hack.
